metrics-report-generator/src/main/scala/it/pagopa/interop/metricsreportgenerator/util/Jobs.scala

```diff
@@ -0,0 +1,126 @@
+  logger: LoggerTakingImplicit[ContextFieldsToLog],
+  context: ContextFieldsToLog
+) {
+
+  def getTokensData(implicit ec: ExecutionContext): Future[List[String]] = {
+
+    logger.info("Gathering tokens information")
+
+    def allTokensPaths(): Future[List[String]] = fileManager.listFiles(config.tokens.bucket)(config.tokens.basePath)
+
+    def getTokens(path: String): Future[List[String]] = fileManager
+      .getFile(config.tokens.bucket)(path)
+      .map(bs => new String(bs).split('\n').toList)
+
+    def getTokenInfo(token: String): Future[(String, String, OffsetDateTime)] = for {
+      fields <- Future(token.parseJson.asJsObject.fields)
+      aId    <- fields
+        .get("agreementId")
+        .collect { case JsString(x) => x }
+        .toFuture(GenericError("Missing agreementId field in token"))
+      pId    <- fields
+        .get("purposeId")
+        .collect { case JsString(x) => x }
+        .toFuture(GenericError("Missing purposeId field in token"))
+      dateL  <- fields
+        .get("issuedAt")
+        .collect { case JsNumber(x) => x.toLong }
+        .toFuture(GenericError("Missing issuedAt field in token"))
+      date   <- dateL.toOffsetDateTime.toFuture
+    } yield (aId, pId, date)
+
+    def allTokenInfo(): Future[List[(String, String, OffsetDateTime)]] = allTokensPaths().flatMap(paths =>
+      Future
+        .traverseWithLatch(config.collections.maxParallelism)(paths)(path =>
+          getTokens(path).flatMap(Future.traverse(_)(getTokenInfo))
+        )
+        .map(_.flatten)
+    )
+
+    allTokenInfo().map(
+      _.map { case (aId, pId, time) => (aId, pId, time.getYear, time.getMonthValue, time.getDayOfMonth) }
+        .groupMapReduce(identity)(_ => 1)(_ + _) // count occurrences
+        .toList
+        .sortBy { case ((aId, pId, y, m, d), _) => (aId, pId, y, m, d) }
+        // to make it prettier you should let me use Shapeless or Scala 3
+        .map { case ((aId, pId, y, m, d), c) => s""""$aId","$pId","$y","$m","$d","$c"""" }
+        .prepended("agreementId,purposeId,year,month,day,tokencount")
+    )
+  }
+
+  def getAgreementRecord(implicit ec: ExecutionContext): Future[List[String]] = {
+    logger.info("Gathering Agreements Information")
+
+    ReadModelQueries
+      .getAllActiveAgreements(config.collections, readModel)
+      .zip(ReadModelQueries.getAllPurposes(config.collections, readModel))
+      .map { case (agreements, purposes) =>
+        val header = "eservice,producer,consumer,activationDate,purposes,agreementId,eserviceId,consumerId,purposeIds"
+        header :: agreements.toList.map { a =>
+          val purposeIds: Seq[(String, String)] = purposes
+            .filter(p => p.consumerId == a.consumerId && p.eserviceId == a.eserviceId)
+            .map(p => (p.purposeId, p.name))
+          List(
+            a.eservice,
+            a.producer,
+            a.consumer,
+            a.activationDate.getOrElse(""),
+            purposeIds.map(_._2).mkString("|"),
```

Comment:
I would quote someone here
`._1 and ._2 are not very readable` 😂

Reply:
besides, this could become a very long list, is it ok to keep it in the same record?

Reply:
besides part 2, there is no guarantee that purpose name does not contain the character `|`